File: regression.py
import numpy as np
from exceptions import RuntimeError

import optimization as opt

class LinearRegression:
    """A basic implementation of linear regression."""
    def __init__(self):
        self.theta_trained = None

    def predict(self,X,theta=None):
        """Predict output based on observations X and linear weights theta."""
        theta = self.theta_trained if theta is None else theta
        if theta is None:
            raise RuntimeError('This model has not been trained.')
        
        return np.dot(X,theta)
    
    def cost(self,X,theta,y,lmbda=0.0):
        """Calculate the cost/objective function given observations X, linear weights theta, and target y."""
        m = X.shape[0]
        return (np.dot(X,theta) - y)**2/(2.0*m) + self.reg_func(theta,lmbda/m)

    def cost_grad(self,X,theta,y,lmbda=0.0):
        """Calculate the gradient of the cost/objective function at theta."""
        m = X.shape[0]
        return np.dot(X.T, np.dot(X,theta) - y)/m + self.reg_grad(theta,lmbda/m)

    def reg_func(self,theta,coeff):
        """Calculate the regularization term (ridge)."""
        return coeff*(np.dot(theta,theta) - theta[0]**2)

    def reg_grad(self,theta,coeff):
        """Calculate the gradient of the regularization term (ridge)."""
        grad = theta.copy()
        grad[0] = 0.0
        return coeff*grad

    def train(self,X,y,lmbda=0.0,tol=0.0001,max_iters=500):
        """Train on data X with target y."""
        def grad_func(theta): return self.cost_grad(X,theta,y,lmbda)
        self.theta_trained,iter = opt.gradient_descent(grad_func,
                                                       0.1,
                                                       np.zeros(X.shape[1]),
                                                       max_iters,
                                                       tol)
        return self.theta_trained

    def solve_normal_eqns(self,X,y,lmbda=0.0):
        """Solve the normal equations to get the linear regression solution."""
        self.theta_trained = np.linalg.solve(np.dot(X.T,X), np.dot(X.T,y))
        return self.theta_trained

File: Regression.jl
module Regression

export LinearRegression, LassoRegression, RidgeRegression
export train!, predict, solve_normal_equations!

using Optimization
using Regularization

abstract Model

"Simple linear regression model"
type LinearRegression <: Model
    theta::Array{Float64,1}
    lmbda::Float64
    alpha::Float64
    penalty_norm::Int
end

LinearRegression(theta,lmbda,alpha) = LinearRegression(theta,lmbda,alpha,0)
LassoRegression(theta,lmbda,alpha) = LinearRegression(theta,lmbda,alpha,1)
RidgeRegression(theta,lmbda,alpha) = LinearRegression(theta,lmbda,alpha,2)

# Cost Functions
# --------------
"""
    cost_lr(model, lmbda, X, y)

Compute the cost function associated with a linear regression model.

Note: the cost function is not used directly in training.  It is provided
for information/validation only.
"""
function cost_lr(theta, lmbda, X, y)
    m = size(X)[1]
    err = X*theta - y
    dot(err,err)/(2*m) + penalty(theta,lmbda)
end


"""
    cost_grad_lr(theta, lmbda, X, y)

Compute the gradient of the cost function associated with a linear regression model.
"""
function cost_grad_lr(theta, lmbda, X, y)
    m = size(X)[1]
    err = X*theta - y
    (X'*err)/m + penalty_grad(theta,lmbda)
end

cost(model::LinearRegression) = cost_lr
cost_grad(model::LinearRegression) = cost_grad_lr


# Training and predicting functions
# ---------------------------------
"""
    train!(model, X, y)

Train model using feature matrix X and labels y.
"""
function train!(model::LinearRegression, X, y; params...)

    cost_model = cost(model)
    cost_grad_model = cost_grad(model)

    objective(theta) = cost_model(theta, model.lmbda, X, y)
    grad(theta) = cost_grad_model(theta, model.lmbda, X, y)

    params = Dict{Any,Any}(params)
    if !(:objective in keys(params))
        params[:objective] = objective
    end
    
    model.theta = gradient_descent(grad, model.alpha, model.theta; params...)
end


"""
    predict(model, X)

Make a prediction on feature matrix X given model.
"""
function predict(model::LinearRegression, X)
    X*model.theta
end


"""
    solve_normal_equations!(model, X, y)

Solve the normal equations for regularized linear regression and
return the parameter vector in model.
"""
function solve_normal_equations!(model::LinearRegression, X, y)
    L = eye(length(model.theta))
    L[1,1] = 0.0
    model.theta = (X'*X + model.lmbda*L)\(X'*y)
end

end
